access/handler.go

```diff
@@ -701,6 +725,392 @@ func (h *Handler) GetExecutionResultByID(ctx context.Context, req *access.GetExe
+	}
+
+	sub := h.api.SubscribeBlockDigests(stream.Context(), flow.ZeroID, 0, blockStatus)
+	return h.handleBlockDigestsSubscription(sub, stream.Send)
+}
+
+// handleBlockDigestsSubscription handles the subscription to block updates and sends
+// the subscribed block digest information to the client via the provided stream.
+//
+// Expected errors during normal operation:
+//   - codes.Internal: if the stream encounters an error, gets an unexpected response,
+//     or cannot convert block to message or send a response.
+func (h *Handler) handleBlockDigestsSubscription(sub subscription.Subscription, send sendSubscribeBlockDigestsResponseFunc) error {
+	for {
+		v, ok := <-sub.Channel()
+		if !ok {
+			if sub.Err() != nil {
+				return rpc.ConvertError(sub.Err(), "stream encountered an error", codes.Internal)
+			}
+			return nil
+		}
+
+		blockDigest, ok := v.(*flow.BlockDigest)
+		if !ok {
+			return status.Errorf(codes.Internal, "unexpected response type: %T", v)
+		}
+
+		err := send(&access.SubscribeBlockDigestsResponse{
+			BlockId:        convert.IdentifierToMessage(blockDigest.ID()),
+			BlockHeight:    blockDigest.Height,
+			BlockTimestamp: timestamppb.New(blockDigest.Timestamp),
+		})
+		if err != nil {
+			return rpc.ConvertError(err, "could not send response", codes.Internal)
+		}
+	}
+}
+
+// getSubscriptionDataFromStartBlockID processes subscription start data from start block id.
+// It takes a union representing the start block id and a BlockStatus from the entities package.
+// Performs validation of input data and returns it in expected format for further processing.
+//
+// Returns:
+// - flow.Identifier: The start block id for searching.
+// - flow.BlockStatus: Block status.
+// - error: An error indicating the result of the operation, if any.
+//
+// Expected errors during normal operation:
+// - codes.InvalidArgument: If blockStatus is flow.BlockStatusUnknown, or startBlockID could not convert to flow.Identifier.
+func (h *Handler) getSubscriptionDataFromStartBlockID(msgBlockId []byte, msgBlockStatus entities.BlockStatus) (flow.Identifier, flow.BlockStatus, error) {
+	startBlockID, err := convert.BlockID(msgBlockId)
+	if err != nil {
+		return flow.ZeroID, flow.BlockStatusUnknown, err
+	}
+
+	blockStatus, err := h.getBlockStatus(msgBlockStatus)
+	if err != nil {
+		return flow.ZeroID, flow.BlockStatusUnknown, err
+	}
+
+	return startBlockID, blockStatus, nil
+}
+
+// getBlockStatus return a block status from the entities package to a flow.BlockStatus and verify block status.
+//
+// Returns:
+// - flow.BlockStatus: Converted block status.
+// - error: An error indicating the result of the conversion, if any.
+//
+// Expected errors during normal operation:
+// - codes.InvalidArgument: If the provided block status is unknown.
+func (h *Handler) getBlockStatus(msgBlockStatus entities.BlockStatus) (flow.BlockStatus, error) {
+	blockStatus := convert.MessageToBlockStatus(msgBlockStatus)
+	err := checkBlockStatus(blockStatus)
```

Comment:
breaking this function out doesn't reduce the amount of code or complexity. let's merge them